device_backends/LogicalNameMapping/src/LNMMathPlugin.cc

```diff
@@ -73,186 +48,151 @@ namespace ChimeraTK::LNMBackend {
-      // empty all queues (initial values, remaining exceptions from previous thread runs). Ignore all exceptions.
-      while(true) {
-        auto notfy = _pushParameterReadGroup.waitAnyNonBlocking();
-        if(!notfy.isReady()) break;
-        try {
-          notfy.accept();
-        }
-        catch(...) {
-          continue;
-        }
-      }
-      assert(!_backend.lock()->_asyncReadActive);
-      assert(_lastWrittenValue.size() == target->getNumberOfSamples());
-
-    } // end scope of the barrierWaitCaller. At this point the finally is executed and other thread is notified that
-    // this thread has started and the opend hook can complete.
-
-    // start processing loop
-    while(true) {
-      try {
-        _pushParameterReadGroup.readAny();
-        {
-          std::unique_lock<std::recursive_mutex> lk(_writeMutex);
-          if(!checkAllParametersWritten(_pushParameterAccessorMap)) {
-            continue;
-          }
-          if(!_mainValueWrittenAfterOpen) {
-            continue;
-          }
+  void MathPluginFormulaHelper::updateResult(ChimeraTK::VersionNumber versionNumber) {
+    assert(_lastMainValue.size() == _target->getNumberOfSamples());
+
+    try {
+      std::unique_lock<std::recursive_mutex> lk(_mp->_writeMutex);
 
-          _h->computeResult(_lastWrittenValue, target->accessChannel(0));
-          target->writeDestructively();
+      auto paramDataValidity = ChimeraTK::DataValidity::ok;
+
+      //  update parameters
+      for(auto& p : params) {
+        p.first->readLatest();
+        // check the DataValidity of parameter.
+        if(p.first->dataValidity() == ChimeraTK::DataValidity::faulty) {
+          // probably compiler optimize it automatically and assign it only once.
+          paramDataValidity = ChimeraTK::DataValidity::faulty;
         }
       }
-      catch([[maybe_unused]] ChimeraTK::runtime_error& e) {
-        // no need to report the exception, as the accessor should have already done it
+
+      if(!checkAllParametersWritten() || !_mp->_mainValueWrittenAfterOpen) {
         return;
       }
+
+      computeResult(_lastMainValue, _target->accessChannel(0));
+      // pass validity to target and delegate preWrite
+      if(paramDataValidity == ChimeraTK::DataValidity::ok && _lastMainValidity == ChimeraTK::DataValidity::ok) {
+        _target->setDataValidity(ChimeraTK::DataValidity::ok);
+      }
+      else {
+        _target->setDataValidity(ChimeraTK::DataValidity::faulty);
+      }
+
+      _target->writeDestructively(versionNumber);
+    }
+    catch(runtime_error&) {
+      // runtime_error from param.readLatest() or target->write()
+      // we could actually even ignore it, since we don't expect exceptions on param.readLatest(), and target->write()
+      // already puts backend into exception state.
+      // TODO discuss - should we direct to lnmBackend->getExceptionBackend() instead?
```

Comment:
@mhier  can you comment on this?

Reply:
I guess _target->getExceptionBackend() will give you LNM backend, but I'm not sure if this in line with the spec. The spec for Exception backend treatment is really hard to get.

Reply:
Yes, you should always report an exception to the exception backend. It might be the logical name mapping backend, or it might be a backend on a higher level (e.g. another LNM backend using the LNM backend which handed out the accessor as a target). Higher level backends are required to pass exceptions down to all their target backends. This way we make sure all participating backends are going into an exception state consistently.

Reply:
Does "yes" mean the code line is fine? More precisely, which is better:
(A)  _target->getExceptionBackend()->setException()
(B) lnmBackend->getExceptionBackend()->setException()

Reply:
(A) is right, (B) does not make sense because getExceptionBanckend() is for accessors